src/celeritas/field/DormandPrinceMultiStepperShared.cuda.hh

```diff
@@ -0,0 +1,363 @@
+    {
+        __syncwarp(mask);
+        for (int j = 0; j <= i; j++)
+        {
+            update_state(
+                index, (*along_state), step * axx[coef_counter], ks[j]);
+            coef_counter++;
+        }
+        __syncwarp(mask);
+    }
+
+    // Vector multiplication for step six: end state
+    __syncwarp(mask);
+    for (int j = 0; j < 6; j++)
+    {
+        if (j == 1)
+            continue;  // because a62 = 0
+        update_state(index, result->end_state, step * axx[coef_counter], ks[j]);
+        coef_counter++;
+    }
+    __syncwarp(mask);
+
+    // Vector mutltiplication for step eight and nine: error and mid state
+    __syncwarp(mask);
+    coef_counter = 0;
+    for (int j = 0; j < 7; j++)
+    {
+        if (j == 1)
+            continue;  // because d72 and c72 = 0
+        update_state(index, result->err_state, step * dxx[coef_counter], ks[j]);
+        update_state(
+            index, result->mid_state, step * cxx[coef_counter] / R(2), ks[j]);
+        coef_counter++;
+    }
+    __syncwarp(mask);
+}
+
+template<class E>
+CELER_FUNCTION void
+DormandPrinceMultiStepperShared<E>::run_sequential(real_type step,
+                                                   OdeState const& beg_state,
+                                                   int id,
+                                                   int mask,
+                                                   OdeState* ks,
+                                                   OdeState* along_state,
+                                                   FieldStepperResult* result) const
+{
+    // First step
+    ks[0] = calc_rhs_(beg_state);
+    *along_state = beg_state;
+    dispatch_vect_mult(mask);
+
+    // Second step
+    ks[1] = calc_rhs_(*along_state);
+    *along_state = beg_state;
+    dispatch_vect_mult(mask);
+
+    // Third step
+    ks[2] = calc_rhs_(*along_state);
+    *along_state = beg_state;
+    dispatch_vect_mult(mask);
+
+    // Fourth step
+    ks[3] = calc_rhs_(*along_state);
+    *along_state = beg_state;
+    dispatch_vect_mult(mask);
+
+    // Fifth step
+    ks[4] = calc_rhs_(*along_state);
+    *along_state = beg_state;
+    dispatch_vect_mult(mask);
+
+    // Sixth step
+    ks[5] = calc_rhs_(*along_state);
+    result->end_state = beg_state;
+    dispatch_vect_mult(mask);
+
+    // Seventh step: the final step
+    ks[6] = calc_rhs_(result->end_state);
+
+    // The error estimate and the mid point
+    result->err_state = {{0, 0, 0}, {0, 0, 0}};
+    result->mid_state = beg_state;
+    dispatch_vect_mult(mask);
+}
+
+template<class E>
+inline CELER_FUNCTION void DormandPrinceMultiStepperShared<E>::update_state(
+    int index, OdeState& state, real_type coefficient, OdeState const& k) const
+{
+    state.pos[index - 1] = coefficient * k.pos[index - 1]
+                           + state.pos[index - 1];
+    state.mom[index - 1] = coefficient * k.mom[index - 1]
+                           + state.mom[index - 1];
```

Comment:
Since these two calculations are independent, couldn't you double the parallelism by having two separate threads perform this calculation too?

Reply:
It is an idee but we see that the requested resources is already too high, I do not think this will speed up the process.